### ISSUE TITLE
Font Info window layout improvment

### ISSUE THREAD

**Issue by pathumego** [trufont/trufont#234]
Current Font info window is organised based on the metadata related with differen specs and formats. 

<img width="874" alt="screen shot 2016-02-06 at 2 01 56 pm" src="https://cloud.githubusercontent.com/assets/6846609/12867030/38136b06-ccda-11e5-9932-3523ea92c095.png">

General, Metrics, OpenType, OS/2, Postscript

Some values in these different tabs repeat and sometimes have some sort of a relationship to each other. So it makes more sense to organise them into tabs based on what metadata is about.

So I suggest to split the tabs in the following order,

- General: Include what is already there, Name Table from Open type tab, Naming section from Postscript Tab, Chrachter section from Postscript
- Matrics : Include what is already there, hhea table and vhea table from open type table, Typo ascend and other matrics related stuff from OS/2, Matrics section from Postscript
- Hinting: Hints section from Postscript

And devide the tabs into sections naming what table that set of metadata refer to would be more clear and intuitive in my opinion. And it will be easier to make the auto filling process.
But I can see some users would prefer current layout. So another option is having two layouts and add this as a preference.

**Comment by madig** [comment]
I'm against adding a preference, that just bloats the code and UI for no gain.

I'd modify your suggestion a bit:
* General: family name, style name, style map, version, date created, name table from OpenType tab, naming section from Postscript tab, character section from Postscript
* Metrics: upM, italic angle, ascender, descender, cap height, x height, hhea/vhea/os2, etc.
* Hinting: hints section from Postscript, possibly the truetype hinting flag stuff
* Legal: designer, manufacturer, copyright, license, trademark
* Note: the note field

**Comment by pathumego** [comment]
@madig Agree :+1:

**Comment by pathumego** [comment]
And,
- I think this descion will decide the basic ideas for our HIG on how we link the specs data to UI. 
- How do we handle any extra data related to Font File? Font Info only houses metadata related to font file and format specification right now. We have to decide how other TruFont specific file data (comes later with more additional functinality and extensions) are handled in the UI. As an extra tabs on Font Info or extra window.

**Comment by madig** [comment]
Hm.
![bildschirmfoto vom 2016-02-06 20-37-03](https://cloud.githubusercontent.com/assets/380829/12868888/6946567a-cd11-11e5-9fa7-7a7e856ced2b.png)

**Comment by pathumego** [comment]
:+1:

**Comment by madig** [comment]
I'm more like :-1: because stuff doesn't align...

**Comment by adrientetar** [comment]
I rather have sorting by table kind since you usually don't fiddle with format-specific naming-data for instance (inferred at compile-time) unless you override the defaults. If I'm going to change the Postscript font name which is different than the other names in the font then I know I'm going to the PS tab.

**Comment by adrientetar** [comment]
Also it makes more sense imo to have e.g. a PS tab because it packs it's own metadata which corresponds to the format. It's redundant with OpenType is most cases so here it gives you overview of the Postscript naming.  
One thing I would like to see is text fields grayed-out with default value inferred at compile-time (you can get that from fontInfoData in ufo2ft/ufo2fdk) displayed, so you can see it and override it if necessary. A while ago I thought about clicking on the label to enable/disable the corresponding edit field.

**Comment by madig** [comment]
Hm, actually I'm okay with that. I assume most people are primarily interested in the current contents of the general and metrics tab. So how about:

* General: 
  * family name, style name, style map, version, date created
  * metrics grouping box: upM, italic angle, ascender, descender, cap height, x height
* OpenType
* OS/2
* PostScript
* Legal: designer, manufacturer, copyright, license, trademark
* Note: the note field

And I like the grey-out-and-display-default thing.

**Comment by davelab6** [comment]
I agree, I was going to suggest a _simple_ tab that allows novices or people who don't care about the details too much and want to be able to say quite simply what the family name is, what style it is, and move on.

**Comment by khaledhosny** [comment]
Additionally, I think entries that can be inferred at build time should be disabled somehow with an option to override them so that people ar not tempted to duplicate the data there. At least way to indicate they can be inferred at build time is a good start.

**Comment by adrientetar** [comment]
I thought about that at some point but to avoid fussy code we'd need automation that loads a label and a line edit for a given attribute, then links the label to the corresponding lineedit for enable/disable. 
Otherwise it could be implemented with additional checkboxes I guess.

**Comment by madig** [comment]
So what's the bare minimum that that people should care about? 
```
requiredAttributes = {'ascender', 'styleName', 'familyName', 'xHeight', 'unitsPerEm', 'descender', 'capHeight'}

recommendedAttributes = set([
    "styleMapFamilyName",
    "versionMajor",
    "versionMinor",
    "copyright",
    "trademark",
    "openTypeHeadCreated",
    "openTypeNameDesigner",
    "openTypeNameDesignerURL",
    "openTypeNameManufacturer",
    "openTypeNameManufacturerURL",
    "openTypeNameLicense",
    "openTypeNameLicenseURL",
    "openTypeNameDescription",
    "openTypeNameSampleText",
    "openTypeOS2WidthClass",
    "openTypeOS2WeightClass",
    "openTypeOS2VendorID",
    "openTypeOS2Panose",
    "openTypeOS2FamilyClass",
    "openTypeOS2UnicodeRanges",
    "openTypeOS2CodePageRanges",
    "openTypeOS2TypoLineGap",
    "openTypeOS2Type",
    "postscriptBlueValues",
    "postscriptOtherBlues",
    "postscriptFamilyBlues",
    "postscriptFamilyOtherBlues",
    "postscriptStemSnapH",
    "postscriptStemSnapV"
])
```
I count: family name, style name, style map?, date created?, version, upM, italic angle, ascender, descender, cap height, x height. These could go into a "General" tab, I think the most important one. The two next, semi-important ones would be "Notes" and "Legal", which would house designer, manufacturer, copyright, license, trademark. The OpenType, OS/2, Postscript Tabs would be mainly grayed-out. What about Postscript hints and some of the OS/2 things?

**Comment by madig** [comment]
How about this?
![bildschirmfoto vom 2016-02-09 18-28-09](https://cloud.githubusercontent.com/assets/380829/12924527/e51163a8-cf5a-11e5-92d2-d20b113e10d4.png)

**Comment by madig** [comment]
Even better :)
![bildschirmfoto vom 2016-02-09 19-35-28](https://cloud.githubusercontent.com/assets/380829/12926555/4e91d8fe-cf64-11e5-9d07-90e48c470c43.png)

**Comment by madig** [comment]
Hm. Compress the version field again?
![bildschirmfoto vom 2016-02-09 19-46-03](https://cloud.githubusercontent.com/assets/380829/12926922/c9fb26fc-cf65-11e5-9931-f857f33d81d4.png)

**Comment by adrientetar** [comment]
I still don't understand how this change is an improvement over the current layout, and how this solves the issue at stake.

> And devide the tabs into sections naming what table that set of metadata refer to would be more clear and intuitive in my opinion.

This is already the case in the OpenType tab (name, hhea, vhea). OS/2 is on a separate tab currently because I haven't worked on adding a scrollbar inside the OpenType tab. And PostScript is just one table, so to speak.

> Some values in these different tabs repeat and sometimes have some sort of a relationship to each other.

My suggestion of greyed-out fields and default values displayed would be much more pertinent than any layout change that breaks a somewhat atomic split into arbitrary recombinations, I reckon.

**Comment by madig** [comment]
My "General" suggestion is mainly an attempt to capture the most interesting properties of a font for the user, as they can be used to derive many other values. The "Metrics" tab can turn into a "Legal" tab then, OpenType/OS2/Postscript would be handled separately as you suggest. I simply find the current contents of "General" and "Metrics" confusingly organized.

**Comment by adrientetar** [comment]
> OpenType/OS2/Postscript would be handled separately as you suggest.

Cool.

> I simply find the current contents of "General" and "Metrics" confusingly organized.

Agreed, it's mainly the Metrics tab though. I put there things that didn't make it into the General tab. Style map things could be in general and note, too I guess. 
In fact maybe Metrics and General should just be merged. Given scrollbars, it seems perfectly doable.

**Comment by anthrotype** [comment]
I haven't followed the whole discussion, but noticed from the screenshots that the font metrics fields seem to allow floats; they should be integers, of course.

**Comment by adrientetar** [comment]
> they should be integers, of course.

It's a mock interface that's shown, the current one is correct.

**Comment by pathumego** [comment]
Agreed. It's just a matter of clarity :)

On Tue, Feb 9, 2016, 9:11 PM Adrien Tétar <notifications@github.com> wrote:

> they should be integers, of course.
>
> It's a mock interface that's shown, the current one is correct.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/trufont/trufont/issues/234#issuecomment-182069156>.
>
-- 

෴ ජය! ෴
පැතුම් එගොඩවත්ත
Pathum Egodawatta
pathum@mooniak.com

**Comment by anthrotype** [comment]
Ok good!

**Comment by pathumego** [comment]
I was agreeing to @madig 's comment on UI :) I have no idea about integer thing

**Comment by madig** [comment]
I made it a double spin box on purpose because the UFO spec says metrics can be integers *or floats*. Will look into scrollbars and unifying OpenType and OS/2.

**Comment by madig** [comment]
Currently making a complete mock-up for the font properties dialog (I find "font info" misleading). I want to include all attributes in `ufoLib.fontInfoAttributesVersion3`.

What to do about the following?
```
 'guidelines',

 'macintoshFONDFamilyID',
 'macintoshFONDName',

 'woffMajorVersion',
 'woffMetadataCopyright',
 'woffMetadataCredits',
 'woffMetadataDescription',
 'woffMetadataExtensions',
 'woffMetadataLicense',
 'woffMetadataLicensee',
 'woffMetadataTrademark',
 'woffMetadataUniqueID',
 'woffMetadataVendor',
 'woffMinorVersion',
```

**Comment by davelab6** [comment]
'Font Info' is used in Fontography, FontLab, FontForge, RoboFont and
Glyphs.

**Comment by madig** [comment]
...oh well.

**Comment by madig** [comment]
Phew, I'm done with the draft. Comments please.
![bildschirmfoto vom 2016-02-11 21-53-51](https://cloud.githubusercontent.com/assets/380829/12990429/a5f2df4c-d10a-11e5-8783-6bce8fd62826.png)
![bildschirmfoto vom 2016-02-11 21-54-05](https://cloud.githubusercontent.com/assets/380829/12990428/a5f2c192-d10a-11e5-89dc-c3ebff647dde.png)
![bildschirmfoto vom 2016-02-11 21-56-05](https://cloud.githubusercontent.com/assets/380829/12990431/a5f3998c-d10a-11e5-8934-8d6cd6d20365.png)
![bildschirmfoto vom 2016-02-11 21-58-08](https://cloud.githubusercontent.com/assets/380829/12990427/a5f0b2bc-d10a-11e5-9d2e-6d85a2993281.png)
![bildschirmfoto vom 2016-02-11 21-58-27](https://cloud.githubusercontent.com/assets/380829/12990430/a5f31674-d10a-11e5-90b2-9c74c76b2e7b.png)

**Comment by madig** [comment]
I still find this collection of labels and fields quite cheap usability-wise, but it's a first step towards covering all attributes. The "Hints" and metrics attributes would probably gain from a graphical presentation.

**Comment by graphicore** [comment]
I suggest you put the checkboxes on the right side of the labels, between label and value field. It will look much more structured and be easier to read. Keep the right alignment of the labels, its good.

**Comment by graphicore** [comment]
Hmm, there are then these boolean fields like `isFixedPitch` and `Force Bold` which would need a different treatment.

**Comment by adrientetar** [comment]
They are tristate for this purpose.

**Comment by graphicore** [comment]
> They are tristate for this purpose.

I don't get it. I made a layout suggestion, nothing that would change the nature of the form fields.

**Comment by adrientetar** [comment]
Force bold and isFixedPitch don't need a checkbox, they are tristate for this purpose. 
I wouldn't add checkboxes anyway but make the labels clickable links to enable/disable the corresponding field.

**Comment by graphicore** [comment]
Yeah, that would work as well.

**Comment by davelab6** [comment]
On 11 February 2016 at 16:58, Adrien Tétar <notifications@github.com> wrote:

> Force bold and isFixedPitch don't need a checkbox, they are tristate for
> this purpose.
> I wouldn't add checkboxes anyway but make the labels clickable links to
> enable/disable the corresponding field
>
This seems unusual to me.

**Comment by graphicore** [comment]
The labels can be clickable and toggle the checkboxes … :-)

**Comment by davelab6** [comment]
Oh I see. That is good.

**Comment by adrientetar** [comment]
> The labels can be clickable and toggle the checkboxes … :-)

Making a case for redundancy? A disabled lineEdit would be grayed out, no need to click-toggle an additional checkbox.

**Comment by davelab6** [comment]
Clicking the label to set the associated checkbox is how GTK and Cocoa work.

**Comment by graphicore** [comment]
> Making a case for redundancy?

Yes! You want to communicate with humans, so more redundancy is actually not a bad idea. It may make it faster to consume.

Note that also in html:

```html
<label>hello <input type="checkbox"/></label>
```

Clicking "Hello" will toggle the checkbox.

**Comment by graphicore** [comment]
> A disabled lineEdit would be grayed out

Do this as well.

**Comment by madig** [comment]
Hm, will look into clickable links. Checking something shouldn't check the bool behind it. Unchecked = implicit default, checked checked = explicit enable, checked unchecked = explicit disable.

**Comment by madig** [comment]
@adrientetar do you have examples where stuff is toggled through links?

**Comment by madig** [comment]
I made something different. How about that? Almost a Robofont copy. Better use of space and less ugly use of checkboxes.
![bildschirmfoto vom 2016-02-12 18-59-08](https://cloud.githubusercontent.com/assets/380829/13015696/dd57ccd0-d1ba-11e5-9fac-ab2faf0f24a2.png)
![bildschirmfoto vom 2016-02-12 18-59-18](https://cloud.githubusercontent.com/assets/380829/13015701/df31c830-d1ba-11e5-88f0-b9acbb44769e.png)
![bildschirmfoto vom 2016-02-12 18-59-38](https://cloud.githubusercontent.com/assets/380829/13015700/df31a42c-d1ba-11e5-8761-cb1ba69ec5b6.png)
![bildschirmfoto vom 2016-02-12 18-59-47](https://cloud.githubusercontent.com/assets/380829/13015702/df331c8a-d1ba-11e5-95d1-d35c75bb2330.png)

**Comment by graphicore** [comment]
Hey, I liked the previous proposal, so here a gimped visualisation of my suggestion. That’s just my two cents, I don't want to push it, just for reference and to give an image.

I think having the checkboxes in one column line like this makes it very easy for the eye to scan the form for activated fields. Of course that would be supported by grayed out inactive fields. It's the most important setting for an optional field and it means the same across the optional fields, having it at a common place is a good move.

![a5f3998c-d10a-11e5-8934-8d6cd6d20365_gimped](https://cloud.githubusercontent.com/assets/393132/13016523/95d411c0-d1bf-11e5-9a19-ef0c8b54cecd.png)

**Comment by madig** [comment]
Hm. I reversed the layout direction for the checkboxes. Not bad either.
![bildschirmfoto vom 2016-02-12 21-01-39](https://cloud.githubusercontent.com/assets/380829/13018757/db3e049e-d1cb-11e5-85cc-30a36fd2c7ea.png)

**Comment by madig** [comment]
Here's the .ui file if anyone wants to play around: https://gist.github.com/madig/b95c4703bce8f2469d9a

**Comment by graphicore** [comment]
:+1:

**Comment by adrientetar** [comment]
Another task is to add ulUnicodeRange and Panose.

**Comment by madig** [comment]
Do they need special handling or what do you mean?

**Comment by adrientetar** [comment]
Add them, they're not there yet.

**Comment by madig** [comment]
Is there an example UFO that uses all attributes? I want to know what lists like `openTypeHeadFlags` look like.

**Comment by madig** [comment]
Okay, found something somewhere in my github grave. Looks like bits are simply collected in a list, so if bits 0, 6 and 8 are toggled, it's represented as `[0, 6, 8]`.

**Comment by madig** [comment]
Hm. The "custom parameter" pane as in https://forum.glyphsapp.com/t/gasp-table-and-other-hinting-settings/2056/5 looks good, too. No long list of crap to wade through and you quickly see all of your custom settings.

**Comment by madig** [comment]
@adrientetar what about the name records table, only support platform 3, encoding ID 1 (Microsoft platform, Unicode)?

**Comment by adrientetar** [comment]
Name records? Which UFO field are we talking of?

**Comment by madig** [comment]
openTypeNameRecords

**Comment by adrientetar** [comment]
Okay, this is new in UFO3 which explains why I missed it. I think we should allow any name records, FontForge has a UI for that. We don't have to do it right now but I think it should have suitable UI. I suspect this is mostly about things like localized names so languageID matters most. 
I agree that since platformID and encodingID are set at compile time for other string fields in the font info it's kind of weird to propose it here. It could make sense to mask it depending on what the intended purpose of openTypeNameRecords is (does anyone need/want to set that differently form the defaults? I can't think of a case where someone would).

**Comment by moyogo** [comment]
@madig nice work. 
Ascender, Descender, Line gap in the OS/2 pane should be called TypoAscender, TypoDescender and TypoLineGap.

@adrientetar 
The more I think about it the more I like FontForge’s name panel or Glyphs’ Custom Parameter input.
In Glyphs, each parameter can have a different kind of input or can open a dialog.

**Comment by davelab6** [comment]
I think there 2 big use cases: the 'just works' case and the 'exact
control' case. FF was poor at the former; a customParameter like table is
good for the latter

**Comment by madig** [comment]
Before I put more work into a design that is to be replaced by something else: Should I go back to the drawing board and take the "custom parameters" route? What's left currently is getting the non-text-fields (ranges, gasp, name records) to work and clean-up/refactoring.

**Comment by davelab6** [comment]
Isn't a cP table additional to what you had?

**Comment by madig** [comment]
I currently try to make all UFO3 attributes accessible in the tabs (except guidelines?). A CPT would make that superfluous.

**Comment by adrientetar** [comment]
Can be done in a followup.

**Comment by davelab6** [comment]
On 26 February 2016 at 08:01, Nikolaus Waxweiler <notifications@github.com>
wrote:

> I currently try to make all UFO3 attributes accessible in the tabs (except
> guidelines?). A CPT would make that superfluous.
>
What about custom values in lib?

**Comment by madig** [comment]
What do you mean? I understand a CPT as a way to access all attributes listed in http://www.unifiedfontobject.org/versions/ufo3/fontinfo.html accessible, which is what I'm doing the hard way now.

**Comment by davelab6** [comment]
I think have explicit UI for everything in the spec is good. I think having
a UI for things not in the spec, eg in lib.plist, is also good.

**Comment by madig** [comment]
Ah, ok. So a CPT would control more than just font.info. Yeah, that can be done later. What does the CPT in Glyphs control?

**Comment by davelab6** [comment]
https://github.com/googlei18n/glyphs2ufo/pull/14#issuecomment-184906546
explains where the list of known CPs is, and if you look at my comments on
an outdated diff I have a longer list

**Comment by madig** [comment]
Fixed in #253.
